Codes/Visualization/ConstructOrderBookVisualization_20200214.R

```diff
@@ -45,41 +61,95 @@ mydata$clocktime<-format(as.POSIXct((mydata$time),origin="1970-01-01",tz="UTC"),
-subt<-sub[which(sub$time<=testtime),]
+##INPUT MATRICES
+cols<-c("size","mpid","price","direction","interval")
+subplot<-data.frame(matrix(NA,0,length(cols))); colnames(subplot)<-cols
+mids<-data.frame(cbind(1:length(min),  rep(NA,length(min)))); colnames(mids)<-c("interval","mid")
+subold<-data.frame(matrix(NA,0,length(Cols))); colnames(subold)<-Cols
+xlabels<-rep(NA,length(min))
+
+for (y in 1:length(min)){
+
+#find state of the order book at closest timestap before testtime
+orbk<-mydata[mydata$time<=testtime,]; orbk<-orbk[nrow(orbk),c(1,8:11)]; orbk$mid<-(orbk$askprice1+orbk$bidprice1)/2
+mids$mid[y]<-orbk$mid
+#find submissions before this interval but after previous interval
+subt<-sub[which(sub$time<testtime&sub$time>=oldtesttime),]
+subt<-rbind(subold,subt) #add orders from previous interval
 #find orders that have been executed or cancelled prior to timestamp
-execanct<-execanc[which(execanc$orderID%in%subt$orderID&execanc$time<=testtime),]
-#sum across executions and cancellations (may be several partial or combo)
-totalect<-aggregate(execanct$size,by=list(Category=execanct$orderID),FUN=sum); colnames(totalect)<-c("orderID","size")
-#discard orders that have been completely executed/cancelled by timestamp
-i<-match(subt$orderID,totalect$orderID)
-subt$execanc<-totalect$size[i]; rm(i) #this is the amount of the submitted order that has been executed or cancelled as of timestamp
-subt$execanc<-replace(subt$execanc,is.na(subt$execanc)==1,0) #replace missing with 0
-subt$remainder<-subt$size-subt$execanc #this is the amount of the order remaining as of the timestamp
-  #q<-which(subt$remainder<0) #sanity check
+execanct<-execanc[which(execanc$orderID%in%subt$orderID&execanc$time<testtime&execanc$time>=oldtesttime),]
+      #sum across executions and cancellations (may be several partial or combo)
+      if (nrow(execanct)!=0){
+      totalect<-aggregate(execanct$size,by=list(Category=execanct$orderID),FUN=sum); colnames(totalect)<-c("orderID","size")
+      #discard orders that have been completely executed/cancelled by timestamp
+      i<-match(subt$orderID,totalect$orderID)
+      subt$execanc<-totalect$size[i]; rm(i) #this is the amount of the submitted order that has been executed or cancelled as of timestamp
+      subt$execanc<-replace(subt$execanc,is.na(subt$execanc)==1,0) #replace missing with 0
+      subt$remainder<-subt$size-subt$execanc #this is the amount of the order remaining as of the timestamp
+      if (length(q<-which(subt$remainder<0))!=0){stop("Negative remainders")} #sanity check
+      rm(execanct,totalect)
+      }
+      else {subt$remainder<-subt$size}
 subt<-subt[subt$remainder!=0,] #these should represent the orders that have not yet been execulated
-rm(execanct,totalect)
+#discard buy orders that are higher than the best bid and sell orders that are lower than the best ask
+subt<-subt[which((subt$direction==1&subt$price<=orbk$bidprice1)|(subt$direction==-1&subt$price>=orbk$askprice1)),]
```

Comment:
I also agree here, and am confused about why these orders exist. I don't think it's because of a bug in my code. These orders can be found in the raw dataset. For example, in the AAPL dataset for 11/04/13, there is a buy submission with order ID #35663013 at 10:13:39.919884 at a price of 523.79. I cannot find any record of this submission being executed or cancelled (i.e., no associated execution or cancellation message with this order ID). Meanwhile, at 10:29:59.306972, the best bid price is listed as being at 523.75, which is lower than the price of order ID #35663013, which (as far as I can tell) should still be on the book. This is happening quite often. For example, between 10:30-10:31, I must remove 342 of 918 orders because they are inside the bid-ask spread.

Reply:
I will double check the data. Thanks.